files/lib/chef_compat/copied_from_chef/chef/provider/apt_repository.rb

```diff
@@ -0,0 +1,266 @@
+
+        raise "The key #{cached_keyfile} is invalid and cannot be used to verify an apt repository." unless key_is_valid?("gpg #{cached_keyfile}", "")
+
+        declare_resource(:execute, "apt-key add #{cached_keyfile}") do
+          sensitive new_resource.sensitive
+          action :run
+          not_if do
+            no_new_keys?(cached_keyfile)
+          end
+          notifies :run, "execute[apt-cache gencaches]", :immediately
+        end
+      end
+
+      def install_key_from_keyserver(key = new_resource.key, keyserver = new_resource.keyserver)
+        cmd = "apt-key adv --recv"
+        cmd << " --keyserver-options http-proxy=#{new_resource.key_proxy}" if new_resource.key_proxy
+        cmd << " --keyserver "
+        cmd << if keyserver.start_with?("hkp://")
+                 keyserver
+               else
+                 "hkp://#{keyserver}:80"
+               end
+
+        cmd << " #{key}"
+
+        declare_resource(:execute, "install-key #{key}") do
+          command cmd
+          sensitive new_resource.sensitive
+          not_if do
+            present = extract_fingerprints_from_cmd("apt-key finger").any? do |fp|
+              fp.end_with? key.upcase
+            end
+            present && key_is_valid?("apt-key list", key.upcase)
+          end
+          notifies :run, "execute[apt-cache gencaches]", :immediately
+        end
+
+        raise "The key #{key} is invalid and cannot be used to verify an apt repository." unless key_is_valid?("apt-key list", key.upcase)
+      end
+
+      def install_ppa_key(owner, repo)
+        url = "https://launchpad.net/api/1.0/~#{owner}/+archive/#{repo}"
+        key_id = Chef::HTTP::Simple.new(url).get("signing_key_fingerprint").delete('"')
+        install_key_from_keyserver(key_id, "keyserver.ubuntu.com")
+      rescue Net::HTTPServerException => e
+        raise "Could not access Launchpad ppa API: #{e.message}"
+      end
+
+      def is_ppa_url?(url)
+        url.start_with?("ppa:")
+      end
+
+      def make_ppa_url(ppa)
+        return unless is_ppa_url?(ppa)
+        owner, repo = ppa[4..-1].split("/")
+        repo ||= "ppa"
+
+        install_ppa_key(owner, repo)
+        "http://ppa.launchpad.net/#{owner}/#{repo}/ubuntu"
+      end
+
+      def build_repo(uri, distribution, components, trusted, arch, add_src = false)
+        uri = make_ppa_url(uri) if is_ppa_url?(uri)
+
+        uri = '"' + uri + '"' unless uri.start_with?("'", '"')
+        components = Array(components).join(" ")
+        options = []
+        options << "arch=#{arch}" if arch
+        options << "trusted=yes" if trusted
+        optstr = unless options.empty?
+                   "[" + options.join(" ") + "]"
+                 end
+        info = [ optstr, uri, distribution, components ].compact.join(" ")
+        repo =  "deb      #{info}\n"
+        repo << "deb-src  #{info}\n" if add_src
+        repo
+      end
+    end
+  end
+end
+
+Chef::Provider::Noop.provides :apt_resource
```

Comment:
yeah, it won't until 12.13.<whatever> is released as a gem. but imma merge this so it gets built and released properly.